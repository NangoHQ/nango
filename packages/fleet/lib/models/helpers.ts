import crypto from 'crypto';

export function generateImage(): string {
    const charset = '0123456789abcdef';
    const length = 40;
    const randomBytes = new Uint8Array(length);
    crypto.getRandomValues(randomBytes);

    const commitHash = Array.from(randomBytes)
        .map((byte) => charset[byte % charset.length])
        .join('');
    return `generated/image:${commitHash}`;
}
