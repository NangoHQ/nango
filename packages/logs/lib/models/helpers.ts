import { nanoid } from '@nangohq/utils';
import type { ConcatOperationList, MessageRow, MessageRowInsert, OperationRow, OperationRowInsert } from '@nangohq/types';
import { z } from 'zod';
import type { estypes } from '@elastic/elasticsearch';
import { defaultOperationExpiration } from '../env.js';
import type { LogContext } from '../client.js';

export const operationIdRegex = z.string().regex(/([0-9]|[a-zA-Z0-9]{20})/);

export interface FormatMessageData {
    account?: { id: number; name: string };
    user?: { id: number } | undefined;
    environment?: { id: number; name: string } | undefined;
    connection?: { id: number; name: string } | undefined;
    integration?: { id: number; name: string; provider: string } | undefined;
    syncConfig?: { id: number; name: string } | undefined; // TODO: rename to script or something similar because it also apply to actions and on-events scripts
    meta?: MessageRow['meta'];
}

export function getFormattedOperation(
    data: OperationRowInsert,
    { account, user, environment, integration, connection, syncConfig, meta }: FormatMessageData = {}
): OperationRow {
    return {
        ...getFormattedMessage(data as unknown as MessageRow),
        message: operationTypeToMessage[`${data.operation.type}:${data.operation.action}` as ConcatOperationList],
        id: data.id || nanoid(),
        operation: data.operation,

        accountId: account?.id ?? data.accountId ?? -1,
        accountName: account?.name || data.accountName || '',

        environmentId: environment?.id ?? data.environmentId ?? null,
        environmentName: environment?.name || data.environmentName || null,

        integrationId: integration?.id ?? data.integrationId ?? null,
        integrationName: integration?.name || data.integrationName || null,
        providerName: integration?.provider || data.providerName || null,

        connectionId: connection?.id ?? data.connectionId ?? null,
        connectionName: connection?.name || data.connectionName || null,

        syncConfigId: syncConfig?.id || data.syncConfigId || null,
        syncConfigName: syncConfig?.name || data.syncConfigName || null,

        jobId: data.jobId || null,
        meta: meta || data.meta || null,

        userId: user?.id || data.userId || null,
        parentId: null
    };
}
export function getFormattedMessage(data: Partial<MessageRow>, { meta }: FormatMessageData = {}): MessageRow {
    const now = new Date();
    return {
        id: data.id || nanoid(), // This ID is for debugging purpose, not for insertion

        source: data.source || 'internal',
        level: data.level || 'info',
        operation: data.operation || null,
        type: data.type || 'log',
        message: data.message || '',
        title: data.title || null,
        code: data.code || null,
        state: data.state || 'waiting',

        accountId: null,
        accountName: null,

        environmentId: null,
        environmentName: null,

        integrationId: null,
        integrationName: null,
        providerName: null,

        connectionId: null,
        connectionName: null,

        syncConfigId: null,
        syncConfigName: null,

        jobId: data.jobId || null,

        userId: null,
        parentId: data.parentId || null,

        error: data.error || null,
        request: data.request || null,
        response: data.response || null,
        meta: meta || data.meta || null,

        createdAt: data.createdAt || now.toISOString(),
        updatedAt: data.updatedAt || now.toISOString(),
        startedAt: data.startedAt || null,
        endedAt: data.endedAt || null,
        expiresAt: data.operation ? data.expiresAt || defaultOperationExpiration.sync() : null
    };
}

// TODO: remove once not used by persist anymore
export const oldLevelToNewLevel = {
    debug: 'debug',
    info: 'info',
    warn: 'warn',
    error: 'error',
    verbose: 'debug',
    silly: 'debug',
    http: 'info'
} as const;

export function getFullIndexName(prefix: string, createdAt: string) {
    return `${prefix}.${new Date(createdAt).toISOString().split('T')[0]}`;
}

export function createCursor({ sort }: estypes.SearchHit): string {
    return Buffer.from(JSON.stringify(sort)).toString('base64');
}

export function parseCursor(str: string): any[] {
    return JSON.parse(Buffer.from(str, 'base64').toString('utf8'));
}

export const operationTypeToMessage: Record<ConcatOperationList, string> = {
    'action:run': 'Action',
    'admin:impersonation': 'Admin logged into another account',
    'auth:create_connection': 'Connection created',
    'auth:post_connection': 'post connection execution',
    'auth:refresh_token': 'Token refreshed',
    'auth:connection_test': 'Connection test',
    'deploy:custom': 'Deploys',
    'deploy:prebuilt': 'Deploys',
    'proxy:call': 'Proxy',
    'sync:cancel': 'Sync execution canceled',
    'sync:init': 'Sync initialized',
    'sync:pause': 'Sync schedule paused',
    'sync:request_run_full': 'Full execution triggered',
    'sync:request_run': 'Incremental execution triggered',
    'sync:run': 'Sync executed',
    'sync:unpause': 'Sync schedule resumed',
    'webhook:incoming': 'External webhook executed',
    'webhook:forward': 'External webhook forwarded',
    'webhook:sync': 'Sync completion webhooks',
    'webhook:connection_create': 'Connection creation webhooks',
    'webhook:connection_refresh': 'Token refresh webhooks',
    'events:post_connection_creation': 'Event-based executions',
    'events:pre_connection_deletion': 'Event-based executions'
};

/**
 * Send buffered logs to elasticsearch
 * Ultimately it would be better to have LogContextBuffer (or an option in LogContext)
 */
export async function flushLogsBuffer(logs: MessageRowInsert[], logCtx: LogContext) {
    await Promise.all(
        logs.map(async (log) => {
            await logCtx.log(log);
        })
    );
}
