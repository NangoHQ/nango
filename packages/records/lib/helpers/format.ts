import md5 from 'md5';
import * as uuid from 'uuid';
import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc.js';
import type { FormattedRecord, UnencryptedRecordData } from '../types.js';
import { Err, Ok } from '@nangohq/utils';
import type { Result } from '@nangohq/utils';

dayjs.extend(utc);

export const formatRecords = ({
    data,
    connectionId,
    model,
    syncId,
    syncJobId,
    softDelete = false
}: {
    data: UnencryptedRecordData[];
    connectionId: number;
    model: string;
    syncId: string;
    syncJobId: number;
    softDelete?: boolean;
}): Result<FormattedRecord[]> => {
    // hashing unique composite key (connection, model, external_id)
    // to generate stable record ids across script executions
    const stableId = (unencryptedData: UnencryptedRecordData): string => {
        const namespace = uuid.v5(`${connectionId}${model}`, uuid.NIL);
        return uuid.v5(`${connectionId}${model}${unencryptedData.id}`, namespace);
    };
    const formattedRecords: FormattedRecord[] = [];
    const now = new Date();
    for (const datum of data) {
        const data_hash = md5(JSON.stringify(datum));

        if (!datum) {
            break;
        }

        if (!datum['id']) {
            const error = new Error(`Missing id field in record: ${JSON.stringify(datum)}. Model: ${model}`);
            return Err(error);
        }

        const formattedRecord: FormattedRecord = {
            id: stableId(datum),
            json: datum,
            // postgresql does not support null bytes in strings
            external_id: String(datum['id']).replaceAll('\x00', ''),
            data_hash,
            model,
            connection_id: connectionId,
            sync_id: syncId,
            sync_job_id: syncJobId
        };

        if (softDelete) {
            const deletedAt = datum['deletedAt'];
            formattedRecord.updated_at = now;
            formattedRecord.deleted_at = deletedAt ? dayjs(deletedAt as string).toDate() : now;
        } else {
            formattedRecord.deleted_at = null;
        }
        formattedRecords.push(formattedRecord);
    }
    return Ok(formattedRecords);
};
