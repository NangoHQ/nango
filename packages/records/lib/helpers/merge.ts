import deepmerge from '@fastify/deepmerge';
import type { UnencryptedRecordData } from '../types';

const merge = deepmerge({
    // define algorithm for merging arrays - merges each item in the array
    // rather than concatenating
    mergeArray: (options) => {
        const clone = options.clone;
        return function (_, source) {
            return clone(source);
        };
    }
});

export function deepMergeRecordData(oldData: UnencryptedRecordData, newData: UnencryptedRecordData): UnencryptedRecordData {
    return merge(oldData, newData) as UnencryptedRecordData;
}
