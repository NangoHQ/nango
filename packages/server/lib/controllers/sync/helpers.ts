export function normalizeSyncParams(syncs: (string | { name: string; variant: string })[]): { syncName: string; syncVariant: string }[] {
    return syncs.map((sync) => {
        if (typeof sync === 'string') {
            if (sync.includes('::')) {
                const [name, variant] = sync.split('::');
                return { syncName: name ?? '', syncVariant: variant ?? '' };
            }
            return { syncName: sync, syncVariant: 'base' };
        }

        return { syncName: sync.name, syncVariant: sync.variant };
    });
}
