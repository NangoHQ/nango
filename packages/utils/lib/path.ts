import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __filename = fileURLToPath(import.meta.url);
export const projectRoot = path.join(__filename, '../../../..');
