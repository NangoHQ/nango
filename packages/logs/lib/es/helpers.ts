import { isTest } from '@nangohq/utils';
import { envs } from '../env.js';
import { logger } from '../utils.js';
import { client } from './client.js';
import { getDailyIndexPipeline, indexMessages, policyRetention } from './schema.js';
import { createMessage } from '../models/messages.js';
import { getFormattedMessage } from '../models/helpers.js';

export async function start() {
    if (!envs.NANGO_LOGS_ENABLED) {
        logger.warning('Elasticsearch is disabled, skipping');
        return;
    }

    logger.info('ðŸ”„ Elasticsearch service starting...');

    await migrateMapping();

    logger.info('âœ… Elasticsearch');
}

export async function migrateMapping() {
    try {
        const index = indexMessages;

        logger.info(`Migrating index "${index.index}"...`);

        // -- Policy
        logger.info(`  Updating policy`);
        await client.ilm.putLifecycle(policyRetention());

        // -- Index
        const existsTemplate = await client.indices.existsIndexTemplate({ name: `${index.index}-template` });
        logger.info(`  ${existsTemplate ? 'updating' : 'creating'} index template "${index.index}"...`);

        await client.indices.putIndexTemplate({
            name: `${index.index}-template`,
            index_patterns: `${index.index}.*`,
            template: {
                settings: index.settings!,
                mappings: index.mappings!,
                aliases: { [index.index]: {} }
            }
        });

        // -- Pipeline
        // Pipeline will automatically create an index based on a field
        // In our case we create a daily index based on "createdAt"
        logger.info(`  Updating pipeline`);
        await client.ingest.putPipeline(getDailyIndexPipeline(index.index));

        const existsAlias = await client.indices.exists({ index: index.index });
        if (!existsAlias) {
            // insert a dummy record to create first index
            logger.info(`  Inserting dummy record`);
            await createMessage(getFormattedMessage({ parentId: '-1' }));
        }
    } catch (err) {
        logger.error(err);
        throw new Error('failed_to_init_elasticsearch');
    }
}

export async function deleteIndex({ prefix }: { prefix: string }) {
    if (!isTest) {
        throw new Error('Trying to delete stuff in prod');
    }

    try {
        const indices = await client.cat.indices({ format: 'json' });
        await Promise.all(
            indices.map(async (index) => {
                if (!index.index?.startsWith(prefix)) {
                    return;
                }

                await client.indices.delete({ index: index.index, ignore_unavailable: true });
            })
        );
    } catch (err) {
        logger.error(err);
        throw new Error('failed_to_deleteIndex');
    }
}
