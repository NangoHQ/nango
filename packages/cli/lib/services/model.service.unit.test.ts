import os from 'node:os';
import fs from 'node:fs';
import path from 'node:path';
import { describe, expect, it } from 'vitest';
import { buildModelsTS, fieldToTypescript, fieldsToTypescript, getExportToJSON } from './model.service.js';
import type { NangoModel } from '@nangohq/types';
import { parse } from './config.service.js';
import { removeVersion } from '../tests/helpers.js';

describe('buildModelTs', () => {
    it('should return empty (with sdk)', () => {
        const res = buildModelsTS({ parsed: { yamlVersion: 'v2', integrations: [], models: new Map() } });
        expect(removeVersion(res)).toMatchSnapshot('');
    });

    it('should output all interfaces', () => {
        const models: NangoModel[] = [
            {
                name: 'Foo',
                fields: [
                    { name: '__string', value: 'string', tsType: true, dynamic: true },
                    { name: 'id', value: 'number', tsType: true }
                ]
            },
            {
                name: 'Bar',
                fields: [
                    { name: 'value', value: null, tsType: true },
                    { name: 'top', value: 'boolean', tsType: true, array: true },
                    { name: 'ref', value: 'Foo', tsType: false, model: true, optional: true },
                    {
                        name: 'union',
                        union: true,
                        value: [
                            { name: '0', value: 'literal1' },
                            { name: '1', value: 'literal2' }
                        ],
                        tsType: false
                    },
                    {
                        name: 'array',
                        array: true,
                        value: [
                            { name: '0', value: 'arr1' },
                            { name: '1', value: 'arr2' }
                        ],
                        tsType: false
                    },
                    {
                        name: 'obj',
                        value: [{ name: 'nes', value: 'ted' }]
                    }
                ]
            },
            {
                name: 'Anonymous_unauthenticated_action_returnType_input',
                fields: [{ name: 'input', value: 'number', tsType: true }],
                isAnon: true
            }
        ];
        const res = buildModelsTS({
            parsed: {
                yamlVersion: 'v2',
                models: new Map(Object.entries(models)),
                integrations: []
            }
        });
        expect(removeVersion(res.split('\n').slice(0, 25).join('\n'))).toMatchSnapshot('');
    });

    it('should support all advanced syntax', () => {
        const parsing = parse(path.resolve(__dirname, `../../fixtures/nango-yaml/v2/advanced-syntax`));
        if (parsing.isErr()) {
            throw parsing.error;
        }

        const res = buildModelsTS({ parsed: parsing.value.parsed! });
        const acc = [];
        for (const line of res.split('\n')) {
            if (line === '// ------ SDK') {
                break;
            }
            acc.push(line);
        }
        expect(removeVersion(acc.join('\n'))).toMatchSnapshot();
    });
});

describe('fieldsToTypescript', () => {
    it.each(['a-b', 'a!b', 'a@', 'a&', 'a#', 'a(', 'a)', 'a%'])('should handle exotic key name', (val) => {
        const res = fieldsToTypescript({ fields: [{ name: val, value: 'string' }] });
        expect(res[0]).toStrictEqual(`  "${val}": 'string';`);
    });
});

describe('fieldToTypescript', () => {
    it('should correctly interpret a string union literal type', () => {
        expect(
            fieldToTypescript({
                field: {
                    name: 'test',
                    union: true,
                    value: [
                        { name: '0', value: 'male' },
                        { name: '1', value: 'female' }
                    ]
                }
            })
        ).toStrictEqual("'male' | 'female'");
    });

    it('should correctly interpret a union literal type with all types', () => {
        expect(
            fieldToTypescript({
                field: {
                    name: 'test',
                    union: true,
                    value: [
                        { name: '0', value: 'male' },
                        { name: '1', value: 'string', tsType: true },
                        { name: '1', value: null, tsType: true },
                        { name: '2', value: 'undefined', tsType: true },
                        { name: '3', value: 1, tsType: true },
                        { name: '4', value: true, tsType: true }
                    ]
                }
            })
        ).toStrictEqual("'male' | string | null | undefined | 1 | true");
    });

    it('should correctly interpret a union literal with models', () => {
        expect(
            fieldToTypescript({
                field: {
                    name: 'test',
                    union: true,
                    value: [
                        { name: '0', value: 'User', model: true },
                        { name: '1', value: 'Account', model: true }
                    ]
                }
            })
        ).toStrictEqual('User | Account');
    });

    it('should correctly interpret a literal array', () => {
        expect(
            fieldToTypescript({
                field: {
                    name: 'test',
                    array: true,
                    value: [
                        { name: '0', value: 'User', model: true },
                        { name: '1', value: 'Account', model: true }
                    ]
                }
            })
        ).toStrictEqual('(User | Account)[]');
    });

    it('should correctly interpret a literal array', () => {
        expect(
            fieldToTypescript({
                field: {
                    name: 'test',
                    union: true,
                    value: [
                        { name: '0', value: 'User', model: true, array: true },
                        { name: '1', value: 'string', tsType: true }
                    ]
                }
            })
        ).toStrictEqual('User[] | string');
    });
});

describe('generate exports', () => {
    describe('json', () => {
        it('should export to JSON', () => {
            const folderTS = path.join(os.tmpdir(), 'cli-exports-json');
            fs.rmSync(folderTS, { recursive: true, force: true });
            fs.mkdirSync(folderTS, { recursive: true });
            const pathTS = path.join(folderTS, 'schema.ts');
            fs.writeFileSync(pathTS, `export interface Test { id: string; name: number[]; }`);

            const res = getExportToJSON({ pathTS });
            expect(removeVersion(JSON.stringify(res, null, 2))).toMatchSnapshot();
        });
    });
});
