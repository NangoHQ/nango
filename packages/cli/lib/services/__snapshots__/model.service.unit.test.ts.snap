// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildModelTs > should output all interfaces 1`] = `
"// ---------------------------
// This file was generated by Nango (vTest)
// It's recommended to version this file
// https://nango.dev
// ---------------------------

// ------ Models
export interface Foo {
  [key: string]: string;
  id: number;
};

export interface Bar {
  value: null;
  top: boolean[];
  ref?: Foo;
  union: 'literal1' | 'literal2';
  array: ('arr1' | 'arr2')[];
  obj: {  nes: 'ted';};
};

/** @deprecated It is recommended to use a Model */
export type Anonymous_unauthenticated_action_returnType_input = number
// ------ /Models
"
`;

exports[`buildModelTs > should return empty (with sdk) 1`] = `
"// ---------------------------
// This file was generated by Nango (vTest)
// It's recommended to version this file
// https://nango.dev
// ---------------------------

// ------ Models

// ------ /Models

// ------ SDK

import type { Nango } from '@nangohq/node';
import type { AxiosInstance, AxiosInterceptorManager, AxiosRequestConfig, AxiosResponse, AxiosError } from 'axios';
import type { ApiEndUser, DBSyncConfig, DBTeam, GetPublicIntegration, RunnerFlags } from '@nangohq/types';

export declare const oldLevelToNewLevel: {
    readonly debug: 'debug';
    readonly info: 'info';
    readonly warn: 'warn';
    readonly error: 'error';
    readonly verbose: 'debug';
    readonly silly: 'debug';
    readonly http: 'info';
};
type LogLevel = 'info' | 'debug' | 'error' | 'warn' | 'http' | 'verbose' | 'silly';
interface Pagination {
    type: string;
    limit?: number;
    response_path?: string;
    limit_name_in_request: string;
    in_body?: boolean;
}
interface CursorPagination extends Pagination {
    cursor_path_in_response: string;
    cursor_name_in_request: string;
}
interface LinkPagination extends Pagination {
    link_rel_in_response_header?: string;
    link_path_in_response_body?: string;
}
interface OffsetPagination extends Pagination {
    offset_name_in_request: string;
    offset_start_value?: number;
    offset_calculation_method?: 'per-page' | 'by-response-size';
}
interface RetryHeaderConfig {
    at?: string;
    after?: string;
}
export interface ProxyConfiguration {
    endpoint: string;
    providerConfigKey?: string;
    connectionId?: string;
    method?: 'GET' | 'POST' | 'PATCH' | 'PUT' | 'DELETE' | 'get' | 'post' | 'patch' | 'put' | 'delete';
    headers?: Record<string, string>;
    params?: string | Record<string, string | number>;
    data?: unknown;
    retries?: number;
    baseUrlOverride?: string;
    paginate?: Partial<CursorPagination> | Partial<LinkPagination> | Partial<OffsetPagination>;
    retryHeader?: RetryHeaderConfig;
    responseType?: 'arraybuffer' | 'blob' | 'document' | 'json' | 'text' | 'stream';
    retryOn?: number[] | null;
}
export interface AuthModes {
    OAuth1: 'OAUTH1';
    OAuth2: 'OAUTH2';
    OAuth2CC: 'OAUTH2_CC';
    Basic: 'BASIC';
    ApiKey: 'API_KEY';
    AppStore: 'APP_STORE';
    Custom: 'CUSTOM';
    App: 'APP';
    None: 'NONE';
    TBA: 'TBA';
    Tableau: 'TABLEAU';
    Jwt: 'JWT';
    Bill: 'BILL';
    TwoStep: 'TWO_STEP';
    Signature: 'SIGNATURE';
}
export type AuthModeType = AuthModes[keyof AuthModes];
interface OAuth1Token {
    oAuthToken: string;
    oAuthTokenSecret: string;
}
interface AppCredentials {
    type: AuthModes['App'];
    access_token: string;
    expires_at?: Date | undefined;
    raw: Record<string, any>;
}
interface AppStoreCredentials {
    type?: AuthModes['AppStore'];
    access_token: string;
    expires_at?: Date | undefined;
    raw: Record<string, any>;
    private_key: string;
}
interface BasicApiCredentials {
    type: AuthModes['Basic'];
    username: string;
    password: string;
}
interface ApiKeyCredentials {
    type: AuthModes['ApiKey'];
    apiKey: string;
}
interface CredentialsCommon<T = Record<string, any>> {
    type: AuthModeType;
    raw: T;
}
interface OAuth2Credentials extends CredentialsCommon {
    type: AuthModes['OAuth2'];
    access_token: string;
    refresh_token?: string;
    expires_at?: Date | undefined;
}
interface OAuth2ClientCredentials extends CredentialsCommon {
    type: AuthModes['OAuth2CC'];
    token: string;
    expires_at?: Date | undefined;
    client_id: string;
    client_secret: string;
}
interface OAuth1Credentials extends CredentialsCommon {
    type: AuthModes['OAuth1'];
    oauth_token: string;
    oauth_token_secret: string;
}
interface TbaCredentials {
    type: AuthModes['TBA'];
    token_id: string;
    token_secret: string;
    config_override: {
        client_id?: string;
        client_secret?: string;
    };
}
interface TableauCredentials extends CredentialsCommon {
    type: AuthModes['Tableau'];
    pat_name: string;
    pat_secret: string;
    content_url?: string;
    token?: string;
    expires_at?: Date | undefined;
}
interface JwtCredentials {
    type: AuthModes['Jwt'];
    privateKeyId?: string;
    issuerId?: string;
    privateKey: {
        id: string;
        secret: string;
    } | string;
    token?: string;
    expires_at?: Date | undefined;
}
interface BillCredentials extends CredentialsCommon {
    type: AuthModes['Bill'];
    username: string;
    password: string;
    organization_id: string;
    dev_key: string;
    session_id?: string;
    user_id?: string;
    expires_at?: Date | undefined;
}
interface TwoStepCredentials extends CredentialsCommon {
    type: AuthModes['TwoStep'];
    [key: string]: any;
    token?: string;
    expires_at?: Date | undefined;
}
interface SignatureCredentials {
    type: AuthModes['Signature'];
    username: string;
    password: string;
    token?: string;
    expires_at?: Date | undefined;
}
interface CustomCredentials extends CredentialsCommon {
    type: AuthModes['Custom'];
}
type UnauthCredentials = Record<string, never>;
type AuthCredentials = OAuth2Credentials | OAuth2ClientCredentials | OAuth1Credentials | BasicApiCredentials | ApiKeyCredentials | AppCredentials | AppStoreCredentials | UnauthCredentials | TbaCredentials | TableauCredentials | JwtCredentials | BillCredentials | TwoStepCredentials | SignatureCredentials | CustomCredentials;
type Metadata = Record<string, unknown>;
interface MetadataChangeResponse {
    metadata: Metadata;
    provider_config_key: string;
    connection_id: string | string[];
}
interface Connection {
    id: number;
    provider_config_key: string;
    connection_id: string;
    connection_config: Record<string, string>;
    created_at: string;
    updated_at: string;
    last_fetched_at: string;
    metadata: Record<string, unknown> | null;
    provider: string;
    errors: {
        type: string;
        log_id: string;
    }[];
    end_user: ApiEndUser | null;
    credentials: AuthCredentials;
}
export declare class ActionError<T = Record<string, unknown>> extends Error {
    type: string;
    payload?: Record<string, unknown>;
    constructor(payload?: T);
}
interface RunArgs {
    sync: string;
    connectionId: string;
    lastSyncDate?: string;
    useServerLastSyncDate?: boolean;
    input?: object;
    metadata?: Metadata;
    autoConfirm: boolean;
    debug: boolean;
    optionalEnvironment?: string;
    optionalProviderConfigKey?: string;
}
export interface NangoProps {
    scriptType: 'sync' | 'action' | 'webhook' | 'on-event';
    host?: string;
    secretKey: string;
    team?: Pick<DBTeam, 'id' | 'name'>;
    connectionId: string;
    environmentId: number;
    environmentName?: string;
    activityLogId?: string | undefined;
    providerConfigKey: string;
    provider: string;
    lastSyncDate?: Date;
    syncId?: string | undefined;
    nangoConnectionId?: number;
    syncJobId?: number | undefined;
    dryRun?: boolean;
    track_deletes?: boolean;
    attributes?: object | undefined;
    logMessages?:
    | {
        counts: {
            updated: number;
            added: number;
            deleted: number;
        };
        messages: unknown[];
    }
    | undefined;
    rawSaveOutput?: Map<string, unknown[]> | undefined;
    rawDeleteOutput?: Map<string, unknown[]> | undefined;
    stubbedMetadata?: Metadata | undefined;
    abortSignal?: AbortSignal;
    syncConfig: DBSyncConfig;
    runnerFlags: RunnerFlags;
    debug: boolean;
    startedAt: Date;
    endUser: {
        id: number;
        endUserId: string | null;
        orgId: string | null;
    } | null;
    axios?: {
        request?: AxiosInterceptorManager<AxiosRequestConfig>;
        response?: {
            onFulfilled: (value: AxiosResponse) => AxiosResponse | Promise<AxiosResponse>;
            onRejected: (value: unknown) => AxiosError | Promise<AxiosError>;
        };
    };
}
export interface EnvironmentVariable {
    name: string;
    value: string;
}
export declare const defaultPersistApi: AxiosInstance;
export declare class NangoAction {
    protected nango: Nango;
    private attributes;
    protected persistApi: AxiosInstance;
    activityLogId?: string | undefined;
    syncId?: string;
    nangoConnectionId?: number;
    environmentId: number;
    environmentName?: string;
    syncJobId?: number;
    dryRun?: boolean;
    abortSignal?: AbortSignal;
    syncConfig?: DBSyncConfig;
    runnerFlags: RunnerFlags;
    connectionId: string;
    providerConfigKey: string;
    provider?: string;
    ActionError: typeof ActionError;
    private memoizedConnections;
    private memoizedIntegration;
    constructor(
        config: NangoProps,
        {
            persistApi
        }?: {
            persistApi: AxiosInstance;
        }
    );
    protected stringify(): string;
    private proxyConfig;
    protected throwIfAborted(): void;
    proxy<T = any>(config: ProxyConfiguration): Promise<AxiosResponse<T>>;
    get<T = any>(config: Omit<ProxyConfiguration, 'method'>): Promise<AxiosResponse<T>>;
    post<T = any>(config: Omit<ProxyConfiguration, 'method'>): Promise<AxiosResponse<T>>;
    put<T = any>(config: Omit<ProxyConfiguration, 'method'>): Promise<AxiosResponse<T>>;
    patch<T = any>(config: Omit<ProxyConfiguration, 'method'>): Promise<AxiosResponse<T>>;
    delete<T = any>(config: Omit<ProxyConfiguration, 'method'>): Promise<AxiosResponse<T>>;
    getToken(): Promise<string | OAuth1Token | OAuth2ClientCredentials | BasicApiCredentials | ApiKeyCredentials | AppCredentials | AppStoreCredentials | UnauthCredentials | CustomCredentials | TbaCredentials | TableauCredentials | JwtCredentials | BillCredentials | TwoStepCredentials | SignatureCredentials>;
    /**
     * Get current integration
     */
    getIntegration(queries?: GetPublicIntegration['Querystring']): Promise<GetPublicIntegration['Success']['data']>;
    getConnection(providerConfigKeyOverride?: string, connectionIdOverride?: string): Promise<Connection>;
    setMetadata(metadata: Metadata): Promise<AxiosResponse<MetadataChangeResponse>>;
    updateMetadata(metadata: Metadata): Promise<AxiosResponse<MetadataChangeResponse>>;
    /**
     * @deprecated please use setMetadata instead.
     */
    setFieldMapping(fieldMapping: Record<string, string>): Promise<AxiosResponse<object>>;
    getMetadata<T = Metadata>(): Promise<T>;
    getWebhookURL(): Promise<string | null | undefined>;
    /**
     * @deprecated please use getMetadata instead.
     */
    getFieldMapping(): Promise<Metadata>;
    /**
     * Log
     * @desc Log a message to the activity log which shows up in the Nango Dashboard
     * note that the last argument can be an object with a level property to specify the log level
     * @example
     * \`\`\`ts
     * await nango.log('This is a log message', { level: 'error' })
     * \`\`\`
     */
    log(
        message: any,
        options?:
            | {
                level?: LogLevel;
            }
            | {
                [key: string]: any;
                level?: never;
            }
    ): Promise<void>;
    log(
        message: string,
        ...args: [
            any,
            {
                level?: LogLevel;
            }
        ]
    ): Promise<void>;
    getEnvironmentVariables(): Promise<EnvironmentVariable[] | null>;
    getFlowAttributes<A = object>(): A | null;
    paginate<T = any>(config: ProxyConfiguration): AsyncGenerator<T[], undefined, void>;
    triggerAction<In = unknown, Out = object>(providerConfigKey: string, connectionId: string, actionName: string, input?: In): Promise<Out>;
    triggerSync(providerConfigKey: string, connectionId: string, syncName: string, fullResync?: boolean): Promise<void | string>;
    private sendLogToPersist;
    private logAPICall;
}
export declare class NangoSync extends NangoAction {
    lastSyncDate?: Date;
    track_deletes: boolean;
    logMessages?:
        | {
            counts: {
                updated: number;
                added: number;
                deleted: number;
            };
            messages: unknown[];
        }
        | undefined;
    rawSaveOutput?: Map<string, unknown[]>;
    rawDeleteOutput?: Map<string, unknown[]>;
    stubbedMetadata?: Metadata | undefined;
    private batchSize;
    constructor(config: NangoProps);
    /**
     * @deprecated please use batchSave
     */
    batchSend<T extends object>(results: T[], model: string): Promise<boolean | null>;
    batchSave<T extends object>(results: T[], model: string): Promise<boolean | null>;
    batchDelete<T extends object>(results: T[], model: string): Promise<boolean | null>;
    batchUpdate<T extends object>(results: T[], model: string): Promise<boolean | null>;
    getMetadata<T = Metadata>(): Promise<T>;
    setMergingStrategy(merging: { strategy: 'ignore_if_modified_after' | 'override' }, model: string): Promise<void>;
    getRecordsByIds<K = string | number, T = any>(ids: K[], model: string): Promise<Map<K, T>>;
}
/**
 * @internal
 *
 * This function will enable tracing on the SDK
 * It has been split from the actual code to avoid making the code too dirty and to easily enable/disable tracing if there is an issue with it
 */
export declare function instrumentSDK(rawNango: NangoAction | NangoSync): NangoAction | NangoSync;
export { };



// ------ /SDK

// ------ Flows
export const NangoFlows = [] as const;
// ------ /Flows
"
`;

exports[`buildModelTs > should support all advanced syntax 1`] = `
"// ---------------------------
// This file was generated by Nango (vTest)
// It's recommended to version this file
// https://nango.dev
// ---------------------------

// ------ Models
export interface Ref {
  id: string;
};

export interface Ref2 {
  name: string;
};

export interface Base {
  unions: string | null | Ref | true[] | Ref2[];
  arrUnion: (Ref | true)[];
  literal: 'hello';
  num: 1;
  literalNull: null[];
  obj: {  nested: string;};
  optional?: string;
  arrayLiteral: any[];
  objectLiteral: Record<string, any>;
  any: any;
  float: number;
  bigint: bigint;
  int: number;
  date: Date;
  char: string;
  undefined: undefined;
  true: true;
  false: false;
  id: string;
  name: string;
};

export interface Record {
  [key: string]: any;
};

export interface RecordUnion {
  [key: string]: any | string | boolean;
};

export interface RecordArrayUnion {
  [key: string]: (any | string)[];
};

export interface RecordObject {
  [key: string]: {  str: string;};
};

/** @deprecated It is recommended to use a Model */
export type Anonymous_unauthenticated_action_inputoutputunion_output = string | null[]

/** @deprecated It is recommended to use a Model */
export type Anonymous_unauthenticated_action_inputoutputunion_input = true | 'hello' | Ref
// ------ /Models
"
`;

exports[`generate exports > json > should export to JSON 1`] = `
"{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "Test": {
      "type": "object",
      "properties": {
        "id": {
          "type": "string"
        },
        "name": {
          "type": "array",
          "items": {
            "type": "number"
          }
        }
      },
      "required": [
        "id",
        "name"
      ],
      "additionalProperties": false
    }
  },
  "$comment": "This file was generated by Nango (vTest)"
}"
`;
